kernel/init.c

```diff
@@ -311,24 +311,19 @@ static int do_device_init(const struct device *dev)
 {
 	int rc = 0;
 
+	/* device initialization has been invoked */
+	dev->state->initialized = true;
```

Comment:
It is not yet obvious to me why the setting of this `initialized` field to true was moved here. Doing so, does raise a potential race condition concern here to me.

The routine `device_is_ready()` checks both `initialized` and `init_res`. If there is a context switch between setting `initialized` to true and a failing call to dev->ops.init(), device_is_ready() might mistakenly think that the device is ready, when it has not yet finished initialization.

Is this possible? I'm not sure, but I wanted to raise this just in case.